### PR TITLE
Prevent dind containers from being targeted by oci-systemd-hooks

### LINKED ISSUE #9599: extended network tests failing due to docker/image/something problems (again)
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_networking/ is currently failing/aborting for *all* builds, apparently due to:

```
Jun 27 14:29:03.964: INFO: At {2016-06-27 14:24:04 -0400 EDT} - event for service-webserver: {kubelet nettest-node-1} FailedSync: Error syncing pod, skipping: failed to "StartContainer" for "POD" with RunContainerError: "runContainer: Error response from daemon: Cannot start container 98da81fa7462f15af1e4406e2e4f6c628f4587d18d0cae34fd28bfaed0933c76: [9] System error: mkdir /sys/fs/cgroup/cpuset/docker: read-only file system"
```

(repeated multiple times and then eventually resulting in `Expected error: "timed out waiting for the condition" not to have occurred`)

@stevekuznetsov @runcom ?

### PR DESCRIPTION
**Pull request by marun**
oci-systemd-hooks detects that a container will run systemd by checking
for an entrypoint command of 'systemd' or 'init' and then mounts most
cgroups read-only to support running systemd in an unprivileged
container.  This is not compatible with dind containers, which require
running with --privileged to ensure cgroups mounted read-write.  This
change hardlinks 'init' to 'dind_init' and uses that as the entrypoint
to avoid having dind containers be targeted by the hook.

I'm including a couple of minor dind fixes with this that came up on the way to the fix.

Should supercede #9613 and fixes #9599.

cc: @openshift/networking @ncdc

### CODE REVIEW

**Comment by marun** [comment]
cc: @mrunalp The hardlinking appears to do the trick.

**Comment by marun** [comment]
[testextended][extended:networking]

**Comment by mrunalp** [comment]
LGTM

File: images/dind/Dockerfile [hunk 1]
```diff
@@ -9,11 +9,22 @@ FROM fedora:23
...
+ docker-storage-setup.service\
+ getty.target\
+ lvm2-lvmetad.service\
+ sys-fs-fuse-connections.mount\
+ systemd-logind.service\
+ systemd-remount-fs.service\
+ systemd-udev-hwdb-update.service\
+ systemd-udev-trigger.service\
+ systemd-udevd.service\
+ systemd-vconsole-setup.service
```
Comment (ncdc) [inline images/dind/Dockerfile]: I see this wasn't here before but I'm guessing you want to add masking this?

**Comment by knobunc** [comment]
LGTM.  Nice work, thanks Maru.

**Comment by ncdc** [comment]
[merge]

File: images/dind/Dockerfile [hunk 2]
```diff
@@ -40,4 +51,10 @@ RUN systemctl enable docker
 
 VOLUME /var/lib/docker
 
-CMD ["/usr/sbin/init"]
+## Hardlink init to another name to avoid having oci-systemd-hooks
+## detect containers using this image as requiring read-only cgroup
+## mounts.  dind containers should be run with --privileged to ensure
+## cgroups mounted with read-write permissions.
+RUN ln /usr/sbin/init /usr/sbin/dind_init
```
Comment (liggitt) [inline images/dind/Dockerfile]: were oci-systemd-hooks doing anything helpful/necessary that this will opt us out of?

Comment (ncdc) [inline images/dind/Dockerfile]: See https://github.com/projectatomic/oci-systemd-hook/#oci-systemd-hooks

Comment (ncdc) [inline images/dind/Dockerfile]: It doesn't look like we would need that

File: images/dind/Dockerfile (hunk 1 above)
Comment (marun) [inline images/dind/Dockerfile]: Yes, I periodically mask out processes that fail to start and that dind doesn't depend on.

**Comment by marun** [comment]
Yum flake #9203

@knobunc @ncdc Please remerge

**Comment by ncdc** [comment]
@marun it's no. 4 in the merge queue. That test failure was a test run, not a merge run.

**Comment by marun** [comment]
@ncdc oops, thank you for the clarification!

**Pull request merged**